cli/src/merge_tools/internal.rs

```diff
@@ -0,0 +1,613 @@
+    let num_bytes: u64 = buf.len().try_into().unwrap();
+    let text = match String::from_utf8(buf) {
+        Ok(text) => {
+            if text.contains('\0') {
+                None
+            } else {
+                Some(text)
+            }
+        }
+        Err(_) => None,
+    };
+    match text {
+        Some(text) => FileContents::Text {
+            contents: text,
+            hash,
+            num_bytes,
+        },
+        None => FileContents::Binary { hash, num_bytes },
+    }
+}
+
+fn read_file_contents(
+    store: &Store,
+    tree: &Tree,
+    path: &RepoPath,
+) -> Result<FileInfo, InternalToolError> {
+    match tree.path_value(path) {
+        None => Ok(FileInfo {
+            file_mode: scm_record::FileMode::absent(),
+            contents: FileContents::Absent,
+        }),
+
+        Some(TreeValue::File { id, executable }) => {
+            let mut reader =
+                store
+                    .read_file(path, &id)
+                    .map_err(|err| InternalToolError::ReadFileBackend {
+                        path: path.clone(),
+                        id: Box::new(id.clone()),
+                        source: err,
+                    })?;
+            let mut buf = Vec::new();
+            reader
+                .read_to_end(&mut buf)
+                .map_err(|err| InternalToolError::ReadFileIo {
+                    path: path.clone(),
+                    id: Box::new(id.clone()),
+                    source: err,
+                })?;
+
+            let file_mode = if executable {
+                scm_record::FileMode(mode::EXECUTABLE)
+            } else {
+                scm_record::FileMode(mode::NORMAL)
+            };
+            let contents = buf_to_file_contents(Some(id.hex()), buf);
+            Ok(FileInfo {
+                file_mode,
+                contents,
+            })
+        }
+
+        Some(TreeValue::Symlink(id)) => {
+            let value =
+                store
+                    .read_symlink(path, &id)
+                    .map_err(|err| InternalToolError::ReadSymlink {
+                        path: path.clone(),
+                        id: Box::new(id.clone()),
+                        source: err,
+                    })?;
+            let file_mode = scm_record::FileMode(mode::SYMLINK);
+            let num_bytes = value.len().try_into().unwrap();
+            Ok(FileInfo {
+                file_mode,
+                contents: FileContents::Text {
+                    contents: value,
+                    hash: Some(id.hex()),
+                    num_bytes,
+                },
+            })
+        }
+
+        Some(TreeValue::Tree(tree_id)) => {
+            unreachable!("list of changed files included a tree: {tree_id:?}");
+        }
+        tree_value @ Some(TreeValue::GitSubmodule(_) | TreeValue::Conflict(_)) => {
+            unimplemented!("internal diff editor not implemented for tree value: {tree_value:?}")
+        }
```

Comment:
What does the existing external difftool do if you try to render a commit with submodules or conflicts? Does it support that?

Reply:
Since it checks out changed files, I *think* submodules are ignored, and conflicts are materialized as diff + snapshot. If we don't have short-time plan to add support for these kinds of tree values, returning a user error would be better than panicking.